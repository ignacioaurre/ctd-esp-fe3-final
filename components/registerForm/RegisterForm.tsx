import React, { FC, useEffect } from 'react'
import Stack from '@mui/material/Stack';
import ControlledTextInput from '../controlledTextInput/controlledTextInput';
import { FormProvider, useForm } from 'react-hook-form';
import *  as yup from 'yup';
import { yupResolver } from '@hookform/resolvers/yup'

const registerSchema = yup.object({
    nombre: yup.string().required("El nombre es requerido"),
    apellido: yup.string().required("El apellido es requerido"),
    email: yup.string().email("Debe ser un email válido").required("El campo es requerido"),
})

type RegisterFormData = {
    nombre: string,
    apellido: string,
    email: string,
}

type RegisterFormProps = {
    title: string,
}

const RegisterForm: FC<RegisterFormProps> = ({title}: RegisterFormProps) => {

    const methods = useForm<RegisterFormData>({
        resolver: yupResolver(registerSchema),
        defaultValues: {
            nombre: "Test",
            apellido: "User",
            email: "test@user.com"
        }
    })

    const { setFocus, handleSubmit } = methods;

    const onSubmit = (data: RegisterFormData) => {
        console.log(data)
    }

    useEffect(() => {
        setFocus("nombre")
    }, [setFocus])

  return (
    <FormProvider {...methods} >
        <form onSubmit={handleSubmit(onSubmit)}>
        <h4>{title}</h4>
        <Stack>
            <ControlledTextInput name="nombre" label="Nombre " />
            <ControlledTextInput name="apellido" label="Apellido " />
            <ControlledTextInput name="email" label="Email " />
        </Stack>
        </form>
    </FormProvider>
  )
}

export default RegisterForm;