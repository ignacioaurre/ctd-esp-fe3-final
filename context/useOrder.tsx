import { OrderContext, OrderContextState } from ".";
import { useContext } from "react";

const useOrder = (): OrderContextState => {
    const context = useContext(OrderContext);
    if (!context) {
        throw new Error(
            'useOrder must be used within a OrderProvider'
        );
    }
    return context;
};

export default useOrder;