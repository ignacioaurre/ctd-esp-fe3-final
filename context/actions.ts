import { Card, Delivery, Register, ComicInfo } from 'dh-marvel/features/Types/state.types';
import { Dispatch } from 'react';
import { Action } from './reducer';

export const selectComic = (dispatch: Dispatch<Action>, comic: ComicInfo) => {
    dispatch({type: 'ADD_COMIC', payload: comic})
}

export const submitRegister = (dispatch: Dispatch<Action>, register: Register) => {
    dispatch({type: 'SUBMIT_REGISTER', payload: register});
}

export const submitDelivery = (dispatch: Dispatch<Action>, delivery: Delivery) => {
    dispatch({type: 'SUBMIT_DELIVERY', payload: delivery});
}

export const submitCard = (dispatch: Dispatch<Action>, card: Card) => {
    dispatch({type: 'SUBMIT_CARD', payload: card});
}

export const submitForm = (dispatch: Dispatch<Action>) => {
    dispatch({ type: 'SUBMIT_FORM'})
};

export const restoreState = (dispatch: Dispatch<Action>) => {
    dispatch({ type: 'RESTORE_STATE' })
}

export const setSnackbar = (dispatch: Dispatch<Action>, snackbar: string) => {
    dispatch({ type: 'SET_SNACKBAR', payload: snackbar})
}