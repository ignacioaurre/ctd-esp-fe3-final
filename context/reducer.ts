import { OrderState } from "context";
import { Comic } from "dh-marvel/features/Types/comic.types"
import { Card, Delivery, Register, Order, ComicInfo } from "dh-marvel/features/Types/state.types"

export const types = {
    ADD_COMIC: 'ADD_COMIC',
    SUBMIT_REGISTER: 'SUBMIT_REGISTER',
    SUBMIT_DELIVERY: 'SUBMIT_DELIVERY',
    SUBMIT_CARD: 'SUBMIT_CARD', 
}

export type Action = 
  |  { type: 'ADD_COMIC' , payload: ComicInfo}
  |  { type: 'SUBMIT_REGISTER', payload: Register }
  |  { type: 'SUBMIT_DELIVERY', payload: Delivery }
  |  { type: 'SUBMIT_CARD', payload: Card };

export const initialState: OrderState = { 
    order: {
        comic: {
            img: "",
            price: 0,
            title: "",
        },
        register: {
            usuario: "",
            apellido: "",
            email: "",
        },
        delivery: {
            direccion: "Siempre Viva 123",
            departamento: "",
            ciudad: "",
            provincia: "",
            codigoPostal: "",
        },
        card: {
            nroTarjeta: "",
            nombreTarjeta: "",
            fechaExp: "",
            codSeguridad: "",
        }
    }
}

export const reducer = (state: OrderState, action: Action) => {
    switch (action.type) {
        case 'ADD_COMIC':
            return {...state, comic: action.payload}
        case 'SUBMIT_REGISTER':
            return {...state, register: action.payload}
        case 'SUBMIT_DELIVERY':
            return {...state, delivery: action.payload}
        case 'SUBMIT_CARD':
            return {...state, card: action.payload}
        default:
      return state;
    }
}