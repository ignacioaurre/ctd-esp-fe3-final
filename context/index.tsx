import { Order } from 'dh-marvel/features/Types/state.types';
import React, { useMemo, useReducer, createContext, FC, Dispatch, PropsWithChildren } from 'react';

import { reducer, initialState, Action } from './reducer';

export interface OrderState {
  order: Order
}

export type OrderContextState = {
  state: {order: Order},
  dispatch: Dispatch<Action>
}

export const OrderContext = createContext<OrderContextState | undefined>(undefined);

const ContextProvider: FC<PropsWithChildren> = ({ children }: PropsWithChildren ) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const contextValue = useMemo(
    () => ({
      state,
      dispatch,
    }),
    [state, dispatch],
  );

  return (
        <OrderContext.Provider value={contextValue} >
          {children}
        </OrderContext.Provider>
      )

};

export default ContextProvider;