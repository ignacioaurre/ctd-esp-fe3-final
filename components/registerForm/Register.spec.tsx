import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import useOrder from 'dh-marvel/context/useOrder';
import ContextProvider from '../../context';
import { initialState } from 'dh-marvel/context/reducer';
import RegisterForm, { registerSchema } from './RegisterForm';
import { Register } from 'dh-marvel/features/Types/state.types';

jest.mock("../../context/useOrder")
const mockUseOrder = useOrder as jest.MockedFunction<typeof useOrder>
const mockDispatch = jest.fn();
mockUseOrder.mockReturnValue({
    state: initialState,
    dispatch: mockDispatch
})
const activeStep = 0;
const handleNext = jest.fn();
const handleBack = jest.fn();
const steps = ['Datos personales', 'Dirección de entrega', 'Sección de pago'];
const props = {
    activeStep,
    handleBack,
    handleNext,
    title: "Datos Personales",
    steps
}

const props2 = {
    activeStep: 2,
    handleBack,
    handleNext,
    title: "Datos Personales",
    steps
}


describe('Component RegisterForm', () => {
    test("Component should render", () => {
    const { baseElement } =
    render(
        <ContextProvider>
            <RegisterForm {...props} />
        </ContextProvider>
    )
    expect(baseElement).toMatchSnapshot();
    })

    test("Should show 3 labels and inputs", () => {
        render(
            <ContextProvider>
                <RegisterForm {...props} />
            </ContextProvider>
        )
        const usernameLabel = screen.getByLabelText('Nombre')
        const usernameInput = screen.getByRole('textbox', {name: /Nombre/i,})
        const lastNameLabel = screen.getByLabelText('Apellido')
        const lastNameInput = screen.getByRole('textbox', {name: /Apellido/i,})
        const emailLabel = screen.getByLabelText('Email')
        const emailInput = screen.getByRole('textbox', {name: /Email/i,})
        expect(usernameInput).toBeInTheDocument();
        expect(usernameLabel).toBeInTheDocument();
        expect(lastNameInput).toBeInTheDocument();
        expect(lastNameLabel).toBeInTheDocument();
        expect(emailInput).toBeInTheDocument();
        expect(emailLabel).toBeInTheDocument();
    })
    test('should render the submit button', () => {
        render(
            <ContextProvider>
                <RegisterForm {...props} />
            </ContextProvider>
        )
        const submitButton = screen.getByRole('button', {name: /Proximo/i,})
        expect(submitButton).toBeInTheDocument();
    });
    test('should have focus on username', () => {
        render(
            <ContextProvider>
                <RegisterForm {...props} />
            </ContextProvider>
        )
        const usernameInput = screen.getByRole('textbox', {name: /Nombre/i,})
        expect(usernameInput).toHaveFocus();
    });

    describe("Validating form", () => {
        test('should return true', async () => {
            const register: Register = {
                nombre: 'valid',
                apellido: 'valid',
                email: 'valid@valid.com',
            }
            expect(await registerSchema.isValid(register)).toBeTruthy();
        })
        test('should return false', async () => {
            const register: Register = {
                nombre: '',
                apellido: 'valid',
                email: 'valid',
            }
            expect(await registerSchema.isValid(register)).toBeFalsy();
        })
        test('should show error message when submit', async () => {
            render(
                <ContextProvider>
                    <RegisterForm {...props} />
                </ContextProvider>
            )
            const emailInput = screen.getByRole('textbox', {name: /Email/i,})
            const submitButton = screen.getByRole('button', {name: /Proximo/i,})
            await userEvent.type(emailInput, "No Email");
            await userEvent.click(submitButton);
            const error = await screen.findByText('Debe ser un email válido');
            expect(error).toBeInTheDocument();
        });
    })

    describe("Submit Form", () => {
        test("Should dispatch info when submited", async () => {
            render(
                <ContextProvider>
                    <RegisterForm {...props} />
                </ContextProvider>
            )
            
            const usernameInput = screen.getByRole('textbox', {name: /Nombre/i,})
            await userEvent.type(usernameInput, "{backspace}{backspace}{backspace}{backspace}")
            await userEvent.type(usernameInput, 'Steve')
            const lastNameInput = screen.getByRole('textbox', {name: /Apellido/i,})
            await userEvent.type(lastNameInput, "{backspace}{backspace}{backspace}{backspace}{backspace}")
            await userEvent.type(lastNameInput, 'Rogers')
            const emailInput = screen.getByRole('textbox', {name: /Email/i,})
            await userEvent.type(emailInput, "{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}{backspace}")
            await userEvent.type(emailInput, 'steve@rogers.com')
            const submitButton = screen.getByRole('button', {name: /Proximo/i,})

            userEvent.click(submitButton);

            await waitFor(() => {
                expect(handleNext).toBeCalled();
            })
            expect(mockDispatch).toBeCalledWith({
                payload: {
                    nombre: "Steve",
                    apellido: "Rogers",
                    email: "steve@rogers.com"
                },
                type: "SUBMIT_REGISTER"
            })
            })
    })

    describe("Last Step", () => {
        test("This is te last step", () => {
            render(
                <ContextProvider>
                    <RegisterForm {...props2} />
                </ContextProvider>
            )

            const submitButton = screen.getByRole('button', {name: /Finalizar/i,})
            expect(submitButton).toBeInTheDocument();
        })
    })
})