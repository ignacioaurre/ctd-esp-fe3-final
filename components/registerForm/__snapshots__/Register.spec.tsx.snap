// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component RegisterForm Component should render 1`] = `
<body>
  <div>
    <form>
      <h4>
        Datos Personales
      </h4>
      <div
        class="css-nen11g-MuiStack-root"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-testid="nombre"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-focused css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r0:"
              id=":r0:-label"
            >
              Nombre 
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-describedby=":r0:-helper-text"
                aria-invalid="true"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="text"
                value="Tony"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1in441m"
                >
                  <span>
                    Nombre 
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-focused MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
              id=":r0:-helper-text"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-testid="apellido"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r1:"
              id=":r1:-label"
            >
              Apellido 
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-describedby=":r1:-helper-text"
                aria-invalid="true"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                type="text"
                value="Stark"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1in441m"
                >
                  <span>
                    Apellido 
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
              id=":r1:-helper-text"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-testid="email"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r2:"
              id=":r2:-label"
            >
              Email 
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-describedby=":r2:-helper-text"
                aria-invalid="true"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r2:"
                type="text"
                value="Tony@stark.com"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1in441m"
                >
                  <span>
                    Email 
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
              id=":r2:-helper-text"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-2royc9"
      >
        <button
          class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-a1sg1c-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Anterior
        </button>
        <div
          class="MuiBox-root css-11qjisw"
        />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Proximo
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</body>
`;
